[Tooling] Allow Installable Build step to fail silently (and still have the step pass) on retries

AliSoftware opened the pull request:
> **Note** This PR depends on https://github.com/wordpress-mobile/release-toolkit/pull/427

## Why?

This PR allows the "Installable Build" steps to not go red on CI when the APK already exists on S3, i.e. were already uploaded from a previous run of the step.

This can happen for example if the CI job failed on some other steps (e.g. UI Tests) but succeeded on the "Installable Build" steps on the first run, and someone uses "Rebuild"—to restart the whole job from scratch, instead of just using Retry on the failed step(s)—which would re-run the whole pipeline including the "Installable Build" steps again. In those cases, we don't want the subsequent re-runs of "Installable Build" steps to fail and go red just because the user started the pipeline to retry other unrelated failures.

## Other

I took the occasion of this PR updating the release-toolkit to ~>6.0 to also get rid of the fastlane/Pluginfile and move the toolkit declaration directly to the Gemfile (as discussed in paaHJt-3BY-p2, since the eval_gemfile dynamic aspect of our Gemfile were breaking Dependabot otherwise)

AliSoftware committed:
🚧 Point to branch of release-toolkit

AliSoftware committed:
Add `skip_if_exists` on Installable Builds

To avoid the case where CI would fail if someone restarts a whole CI pipeline from scratch and the Installable Build step would run a second time after having already uploaded the APK the first time around.

AliSoftware committed:
🚧 Add a force-crash just to test the case of retries

AliSoftware commented on fastlane/Pluginfile:
```diff
@@ -6,5 +6,5 @@ group :screenshots, optional: true do
 
-# gem 'fastlane-plugin-wpmreleasetoolkit', git: 'git@github.com:wordpress-mobile/release-toolkit.git', branch: 'trunk'
-gem 'fastlane-plugin-wpmreleasetoolkit', '~> 5.6'
+gem 'fastlane-plugin-wpmreleasetoolkit', git: 'git@github.com:wordpress-mobile/release-toolkit.git', branch: 'upload_to_s3-skip_if_exists'
+# gem 'fastlane-plugin-wpmreleasetoolkit', '~> 5.6'
```
To be switched back to a stable version when https://github.com/wordpress-mobile/release-toolkit/pull/427 lands and we release a new version of the toolkit.

AliSoftware commented on fastlane/lanes/build.rb:
```diff
@@ -287,6 +288,8 @@ def upload_installable_build(product:)
       message = "#{product} Installable Build: [#{filename}](#{install_url})"
       sh('buildkite-agent', 'annotate', message, '--style', 'info', '--context', "installable-build-#{product}")
     end
+
+    UI.user_error!('🚧 Force crash just to force the CI step to fail so that we can hit retry on it and test the skip_if_exists works as expected')
```
To remove once we've validated that retrying the step after it failed (due to this `user_error!`, so after the first upload to S3 succeeded) will lead to the `upload_to_s3` just log the error.

So:
 - On first run, we expect `upload_to_s3` to succeed, log `"Upload Complete"`, and fail on this ad-hoc `user_error!`
 - On second run, we expect `upload_to_s3` to skip the upload, logging `"File already exists in S3 bucket"`, not crash, and still fail but only on this ad-hoc `user_error!` (and not earlier like before)

AliSoftware replied:
Proof that it worked as expected:

1. [First run](https://buildkite.com/automattic/wordpress-android/builds/7497#01845cb0-8ce7-493d-b2af-d46857271bda) contains those logs:

```
[2022-11-09T14:18:51Z] [14:18:51]: ----------------------------
[2022-11-09T14:18:51Z] [14:18:51]: --- Step: upload_to_s3 ---
[2022-11-09T14:18:51Z] [14:18:51]: --------------------------
[2022-11-09T14:18:51Z] [14:18:51]: Uploading /var/lib/buildkite-agent/builds/ci-android-i-060d21c67755f117e-1/automattic/wordpress-android/WordPress/build/outputs/apk/wordpressJalapeno/debug/org.wordpress.android-wordpress-jalapeno-debug.apk to: ea0ffdd4a98eeacfb5a47250241399da8490ef4e/wordpress-installable-build-pr17442-3e5a495.apk
[2022-11-09T14:18:53Z] [14:18:53]: Upload Complete
```

2. [Second run, after hitting Retry](https://buildkite.com/automattic/wordpress-android/builds/7497#01845cd0-03c7-4f0b-9c56-5a916043594b), contains those logs:

```
[2022-11-09T14:47:33Z] [14:47:33]: --------------------------
[2022-11-09T14:47:33Z] [14:47:33]: --- Step: upload_to_s3 ---
[2022-11-09T14:47:33Z] [14:47:33]: --------------------------
[2022-11-09T14:47:33Z] [14:47:33]: File already exists in S3 bucket a8c-apps-public-artifacts at ea0ffdd4a98eeacfb5a47250241399da8490ef4e/wordpress-installable-build-pr17442-3e5a495.apk. Skipping upload.
```

AliSoftware commented on .buildkite/commands/installable-build.sh:
```diff
@@ -1,5 +1,9 @@
 #!/bin/bash -eu
 
+echo "--- :github: Add GitHub to Known Hosts"
+# to allow pointing gems to git repos and branches, e.g. directly while working on release-toolkit changes
+add_host_to_ssh_known_hosts github.com
```
EDIT: This doesn't seems to be enough to solve the issue to allow `gem 'xxx', git: 'git@github.com:…', …` references in the `fastlane/Pluginfile` to work, as it then fails on rejected `public_key` authentication.

So for now the solution to allow the `Pluginfile` to temporarily point to a branch of the toolkit ended up being to use a `https://github.com/` URL instead of a `git@github.com:` one. Didn't bother to remove it now since I'll remove both this and the pointing to `https://` URL when we'll release a new version of `release-toolkit` after https://github.com/wordpress-mobile/release-toolkit/pull/427 lands and then point back to the official RubyGems reference, but just a reminder to self to remove it before merging.

oguzkocer reviewed (approved):
The changes in this repo looks good to me and the `release-toolkit` PR seems to be already approved :shipit:

AliSoftware committed:
Switch to release-toolkit 6.0.0

And get rid of `fastlane/Pluginfile` to move all the gems declarations in the main `Gemfile.lock`

AliSoftware committed:
Revert "🚧 Add a force-crash just to test the case of retries"

This reverts commit 7f399d25816597be2788f77a81c4147a6f600924.
